kasa/kasacamprotocol.py

```diff
@@ -0,0 +1,132 @@
+"""Protocol implementation for controlling TP-Link Kasa Cam devices."""
+
+import dataclasses
+import errno
+import logging
+from base64 import b64decode, b64encode
+from dataclasses import dataclass
+from pprint import pformat as pf
+from typing import Dict
+from urllib.parse import quote
+
+import httpx
+
+# When support for cpython older than 3.11 is dropped
+# async_timeout can be replaced with asyncio.timeout
+from async_timeout import timeout as asyncio_timeout
+
+from .credentials import Credentials
+from .exceptions import SmartDeviceException
+from .json import loads as json_loads
+from .protocol import TPLinkSmartHomeProtocol
+
+_LOGGER = logging.getLogger(__name__)
+_NO_RETRY_ERRORS = {errno.EHOSTDOWN, errno.EHOSTUNREACH, errno.ECONNREFUSED}
+
+
+@dataclass
+class KasaCamCredentials(Credentials):
+    """Adapter to store credentials in format expected by Kasa Cam devices."""
+
+    def __post_init__(self):
+        """Base64 encode the password as expected by the Kasa Cam protocol."""
+        self.password: str = b64encode(self.password.encode())
+
+    @classmethod
+    def from_instance(cls, instance):
+        """Create KasaCamCredentials from Credentials."""
+        return cls(**dataclasses.asdict(instance))
+
+
+class KasaCamProtocol(TPLinkSmartHomeProtocol):
+    """Implementation of the Kasa Cam protocol."""
+
+    DEFAULT_PORT = 10443
+
+    def __init__(self, *args, **kwargs) -> None:
+        self.credentials = kwargs.pop("credentials")
+        if not isinstance(self.credentials, KasaCamCredentials):
+            self.credentials = KasaCamCredentials.from_instance(self.credentials)
+
+        super().__init__(*args, **kwargs)
+
+        self.session: httpx.Client = None
+
+    async def _connect(self, timeout: int) -> None:
+        if self.session:
+            return
+
+        self.session = httpx.Client(
+            base_url=f"https://{self.host}:{self.port}",
+            auth=httpx.BasicAuth(self.credentials.username, self.credentials.password),
```

Comment:
See my note about performing base64-encoding here instead of creating a new class for it.

Reply:
fixed